### PR TITLE
8323795: jcmd Compiler.codecache should print total size of code cache

### PR DESCRIPTION
**Pull request by y1yang0**
```
CodeHeap 'non-profiled nmethods': size=118592Kb used=29Kb max_used=29Kb free=118562Kb
 bounds [0x00007fbe84622000, 0x00007fbe84892000, 0x00007fbe8b9f2000]
CodeHeap 'profiled nmethods': size=118588Kb used=80Kb max_used=80Kb free=118507Kb
 bounds [0x00007fbe7c9f2000, 0x00007fbe7cc62000, 0x00007fbe83dc1000]
CodeHeap 'non-nmethods': size=8580Kb used=1258Kb max_used=1834Kb free=7321Kb
 bounds [0x00007fbe83dc1000, 0x00007fbe84031000, 0x00007fbe84622000]
 total_blobs=474 nmethods=87 adapters=293
 compilation: enabled
              stopped_count=0, restarted_count=0
 full_count=0
```

It's better to accumulates total size of used/free/size, for example
```
-SegmentedCodeCache
CodeCache: size=245760Kb used=1366Kb max_used=1943Kb free=244393Kb
 bounds [0x00007fdcc89f2000, 0x00007fdcc8c62000, 0x00007fdcd79f2000]
 total_blobs=474, nmethods=87, adapters=293
 stopped_count=0, restarted_count=0, full_count=0 
 compilation=enabled
```

```
+SegmentedCodeCache
CodeHeap 'non-profiled nmethods': size=118592Kb used=29Kb max_used=29Kb free=118562Kb
 bounds [0x00007f89c8622000, 0x00007f89c8892000, 0x00007f89cf9f2000]
CodeHeap 'profiled nmethods': size=118588Kb used=80Kb max_used=80Kb free=118507Kb
 bounds [0x00007f89c09f2000, 0x00007f89c0c62000, 0x00007f89c7dc1000]
CodeHeap 'non-nmethods': size=8580Kb used=1258Kb max_used=1834Kb free=7321Kb
 bounds [0x00007f89c7dc1000, 0x00007f89c8031000, 0x00007f89c8622000]
CodeCache: size=245760Kb, used=1367Kb, max_used=1943Kb, free=244390Kb
 total_blobs=474, nmethods=87, adapters=293
 stopped_count=0, restarted_count=0, full_count=0 
 compilation=enabled
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8323795](https://bugs.openjdk.org/browse/JDK-8323795): jcmd Compiler.codecache should print total size of code cache (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17445/head:pull/17445` \
`$ git checkout pull/17445`

Update a local copy of the PR: \
`$ git checkout pull/17445` \
`$ git pull https://git.openjdk.org/jdk.git pull/17445/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17445`

View PR using the GUI difftool: \
`$ git pr show -t 17445`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17445.diff">https://git.openjdk.org/jdk/pull/17445.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17445#issuecomment-1893537680)

### CODE REVIEW

**Comment by vnkozlov** [comment]
Can you do next layout of output?:
```
CodeHeap 'non-nmethods': size=8580Kb used=1258Kb max_used=1834Kb free=7321Kb
 bounds [0x00007fbe83dc1000, 0x00007fbe84031000, 0x00007fbe84622000]
Total CodeHeap: size=245760Kb, used=1367Kb, max used=1943Kb, free=244390Kb
 total_blobs=474 nmethods=87 adapters=293, full_count=0
Compilation: enabled,  stopped_count=0, restarted_count=0
```

File: src/hotspot/share/code/codeCache.cpp [hunk 1]
```diff
@@ -1772,17 +1783,23 @@ void CodeCache::print_summary(outputStream* st, bool detailed) {
...
+    st->print_cr(" stopped_count=%d, restarted_count=%d, full_count=%d",
                  CompileBroker::get_total_compiler_stopped_count(),
-                 CompileBroker::get_total_compiler_restarted_count());
-    st->print_cr(" full_count=%d", full_count);
+                 CompileBroker::get_total_compiler_restarted_count(),
+                 full_count);
+    st->print_cr(" compilation=%s", CompileBroker::should_compile_new_jobs() ?
+              "enabled" : Arguments::mode() == Arguments::_int ?
+              "disabled (interpreter mode)" :
+              "disabled (not enough contiguous free space left)");
```
Comment (TobiHartmann) [inline src/hotspot/share/code/codeCache.cpp:1802]: Why did you change the order of the `compilation=` and the `stopped_count=` output?

**Comment by TobiHartmann** [comment]
The title is confusing. Should it be something like "jcmd Compiler.codecache should print total size of code cache"?

**Comment by y1yang0** [comment]
@TobiHartmann @vnkozlov 

In the first commit, the ouput is confusing(CodeCache and Total CodeHeap) if SegmentedCodeCache is turned off, i.e. 
```
CodeCache: size=118592Kb used=29Kb max_used=29Kb free=118562Kb
 bounds [0x00007fbe84622000, 0x00007fbe84892000, 0x00007fbe8b9f2000]
Total CodeHeap: size=118592Kb used=29Kb max_used=29Kb free=118562Kb
 total_blobs=474 nmethods=87 adapters=293
 compilation: enabled
              stopped_count=0, restarted_count=0
 full_count=0
```

So I made new change to 
```
-SegmentedCodeCache
CodeCache: size=245760Kb used=1366Kb max_used=1943Kb free=244393Kb
 bounds [0x00007fdcc89f2000, 0x00007fdcc8c62000, 0x00007fdcd79f2000]
 total_blobs=474, nmethods=87, adapters=293
 stopped_count=0, restarted_count=0, full_count=0 
 compilation=enabled

+SegmentedCodeCache
CodeHeap 'non-profiled nmethods': size=118592Kb used=29Kb max_used=29Kb free=118562Kb
 bounds [0x00007f89c8622000, 0x00007f89c8892000, 0x00007f89cf9f2000]
CodeHeap 'profiled nmethods': size=118588Kb used=80Kb max_used=80Kb free=118507Kb
 bounds [0x00007f89c09f2000, 0x00007f89c0c62000, 0x00007f89c7dc1000]
CodeHeap 'non-nmethods': size=8580Kb used=1258Kb max_used=1834Kb free=7321Kb
 bounds [0x00007f89c7dc1000, 0x00007f89c8031000, 0x00007f89c8622000]
CodeCache: size=245760Kb, used=1367Kb, max_used=1943Kb, free=244390Kb
 total_blobs=474, nmethods=87, adapters=293
 stopped_count=0, restarted_count=0, full_count=0 
 compilation=enabled
```

**Comment by vnkozlov** [comment]
`stopped_count` and `restarted_count` are related to compilation. That is why I suggested to print them on the same line:
```
Compilation: enabled, stopped_count=0, restarted_count=0
```

`full_count` can be put on the line with numbers for methods and others.

I am fine with last refactoring you did for different `SegmentedCodeCache` flag values.

**Comment by TobiHartmann** [comment]
I agree with Vladimir, I'd prefer something like this:

```
CodeHeap 'non-nmethods': size=8580Kb used=1258Kb max_used=1834Kb free=7321Kb
 bounds [0x00007f89c7dc1000, 0x00007f89c8031000, 0x00007f89c8622000]
CodeCache: size=245760Kb, used=1367Kb, max_used=1943Kb, free=244390Kb
 total_blobs=474, nmethods=87, adapters=293, full_count=0

Compilation: enabled, stopped_count=0, restarted_count=0
```

**Comment by y1yang0** [comment]
Now it looks like
```
104640:
CodeHeap 'non-profiled nmethods': size=118592Kb used=29Kb max_used=29Kb free=118562Kb
 bounds [0x00007f09f8622000, 0x00007f09f8892000, 0x00007f09ff9f2000]
CodeHeap 'profiled nmethods': size=118588Kb used=80Kb max_used=80Kb free=118507Kb
 bounds [0x00007f09f09f2000, 0x00007f09f0c62000, 0x00007f09f7dc1000]
CodeHeap 'non-nmethods': size=8580Kb used=1257Kb max_used=1833Kb free=7323Kb
 bounds [0x00007f09f7dc1000, 0x00007f09f8031000, 0x00007f09f8622000]
CodeCache: size=245760Kb, used=1366Kb, max_used=1942Kb, free=244392Kb
 total_blobs=474, nmethods=87, adapters=293, full_count=0
Compilation: enabled, stopped_count=0, restarted_count=0

110115:
CodeCache: size=245760Kb used=1366Kb max_used=1935Kb free=244393Kb
 bounds [0x00007ff4d89f2000, 0x00007ff4d8c62000, 0x00007ff4e79f2000]
 total_blobs=474, nmethods=87, adapters=293, full_count=0
Compilation: enabled, stopped_count=0, restarted_count=0
```

File: src/hotspot/share/code/codeCache.cpp [hunk 2]
```diff
@@ -1772,17 +1783,22 @@ void CodeCache::print_summary(outputStream* st, bool detailed) {
...
+    st->print_cr(" total_blobs=" UINT32_FORMAT ", nmethods=" UINT32_FORMAT
+                 ", adapters=" UINT32_FORMAT ", full_count=" UINT32_FORMAT,
+                 blob_count(), nmethod_count(), adapter_count(), full_count);
+    st->print_cr("Compilation: %s, stopped_count=%d, restarted_count=%d",
+                CompileBroker::should_compile_new_jobs() ?
+                "enabled" : Arguments::mode() == Arguments::_int ?
+                "disabled (interpreter mode)" :
+                "disabled (not enough contiguous free space left)",
+                CompileBroker::get_total_compiler_stopped_count(),
+                CompileBroker::get_total_compiler_restarted_count());
```
Comment (TobiHartmann) [inline src/hotspot/share/code/codeCache.cpp:1801]: The indentation should be fixed.

**Review by TobiHartmann — APPROVED** [review]
Thanks, that looks good to me. I'll run testing and report back once it passed.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
`/test/hotspot/jtreg/vmTestbase/vm/compiler/CodeCacheInfo/Test.java fails`

```
 stdout: [CodeHeap 'non-profiled nmethods': size=120032Kb used=3Kb max_used=3Kb free=120028Kb
 bounds [0x0000000113a84000, 0x0000000113cf4000, 0x000000011afbc000]
CodeHeap 'profiled nmethods': size=120016Kb used=14Kb max_used=14Kb free=120001Kb
 bounds [0x000000010bfbc000, 0x000000010c22c000, 0x00000001134f0000]
CodeHeap 'non-nmethods': size=5712Kb used=1752Kb max_used=1752Kb free=3959Kb
 bounds [0x00000001134f0000, 0x0000000113760000, 0x0000000113a84000]
CodeCache: size=245760Kb, used=1769Kb, max_used=1769Kb, free=243988Kb
 total_blobs=748, nmethods=15, adapters=651, full_count=0
Compilation: enabled, stopped_count=0, restarted_count=0
];
 stderr: [java version "23-internal" 2024-09-17
Java(TM) SE Runtime Environment (fastdebug build 23-internal-2024-01-29-1002351.tobias.hartmann.jdk2)
Java HotSpot(TM) 64-Bit Server VM (fastdebug build 23-internal-2024-01-29-1002351.tobias.hartmann.jdk2, mixed mode, sharing)
]
 exitValue = 0

java.lang.RuntimeException: '^(CodeHeap '[^']+': size=\\d+Kb used=\\d+Kb max_used=\\d+Kb free=\\d+Kb\\n bounds \\[0x[0-9a-f]+, 0x[0-9a-f]+, 0x[0-9a-f]+\\]\\n)+ total_blobs=\\d+ nmethods=\\d+ adapters=\\d+\\n compilation: enabled\\n' missing from stdout
	at jdk.test.lib.process.OutputAnalyzer.stdoutShouldMatch(OutputAnalyzer.java:389)
	at vm.compiler.CodeCacheInfo.Test.main(Test.java:78)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1575)
```

**Review by vnkozlov — APPROVED** [review]
The output looks fine to me now. I approve but it needs new round of testing.

**Comment by TobiHartmann** [comment]
I re-submitted testing and will report back once it passed.

**Review by TobiHartmann — APPROVED** [review]
Looks good. I have the fix for the internal test ready.

**Comment by TobiHartmann** [comment]
Please do an integrate delegate (https://wiki.openjdk.org/display/SKARA/Pull+Request+Commands#PullRequestCommands-/integrate), so that I can push both changes at the same time. Thanks!

**Comment by y1yang0** [comment]
Thanks for reviews @vnkozlov @TobiHartmann 

/integrate delegate

**Pull request closed without merging**